Added squadcast terraformer

mohammadmuazam opened the pull request:
Added Squadcast Terraformer with following resources:
- [squadcast_user](https://registry.terraform.io/providers/SquadcastHub/squadcast/latest/docs/resources/user)
- [squadcast_squad](https://registry.terraform.io/providers/SquadcastHub/squadcast/latest/docs/resources/squad)
- [squadcast_team](https://registry.terraform.io/providers/SquadcastHub/squadcast/latest/docs/resources/team)
- [squadcast_team_member](https://registry.terraform.io/providers/SquadcastHub/squadcast/latest/docs/resources/team_member)
- [squadcast_team_roles](https://registry.terraform.io/providers/SquadcastHub/squadcast/latest/docs/resources/team_roles)
- [squadcast_escalation_policy](https://registry.terraform.io/providers/SquadcastHub/squadcast/latest/docs/resources/escalation_policy)
- [squadcast_service](https://registry.terraform.io/providers/SquadcastHub/squadcast/latest/docs/resources/service)
- [squadcast_runbook](https://registry.terraform.io/providers/SquadcastHub/squadcast/latest/docs/resources/runbook)
- [squadcast_slo](https://registry.terraform.io/providers/SquadcastHub/squadcast/latest/docs/resources/slo)

roshan8 committed:
feat: Added user resource to squadcast provider

roshan8 committed:
fix: Api url

roshan8 committed:
refactor

Lucifer8729 committed:
add refresh-token, region, team-name flag

Commit:
fix: updated logic to get host

Lucifer8729 committed:
feat: added team resource to squadcast provider

Lucifer8729 committed:
fix: added logic for different region

roshan8 committed:
fix: Pass region env variable to terraform provider

Lucifer8729 committed:
fix: user service for eu is fixed

Lucifer8729 committed:
feat: added team roles and escalation policy to squadcast provider

Lucifer8729 committed:
feat: added team member resource to squadcast provider

Lucifer8729 committed:
feat: added service resource to squadcast provider

Lucifer8729 committed:
fix: overriding refresh token through cli

Commit:
feat: Added all resources and refactored code

Commit:
Added Squadcast docs

Commit:
Added Squadcast to README

Commit:
Added Squadcast docs

Commit:
Added Squadcast docs

Commit:
updated Squadcast docs

Commit:
Added squadcast provider

Commit:
added new branch

Commit:
Added support for custom server url

Commit:
fix: removed api-endpoint for now

Commit:
Merge branch 'master' into squadcast-terraformer

roshan8 commented on cmd/provider_cmd_squadcast.go:
```diff
@@ -0,0 +1,45 @@
+		Long:  "Import current state to Terraform configuration from Squadcast",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			originalPathPattern := options.PathPattern
+			for _, region := range options.Regions {
```
we don't need to loop over the region. refer to other providers for reference(`vault`)
Though we have multiple dc, app instances are separately hosted.

mohammadmuazam replied:
Initially we used this approach because the argument of `newCmdSquadcastImporter` had an options argument which had a Region parameter ([]string{}), but now we switched it to a variable (string)

roshan8 commented on cmd/provider_cmd_squadcast.go:
```diff
@@ -0,0 +1,45 @@
+			originalPathPattern := options.PathPattern
+			for _, region := range options.Regions {
+				provider := newSquadcastProvider()
+				options.PathPattern = originalPathPattern
+				options.PathPattern += region + "/"
+				log.Println(provider.GetName() + " importing region " + region)
```
mostly not required

mohammadmuazam replied:
Specifying PathPattern is necessary for creating specific folder (us/ or eu/) in the generated folder

roshan8 replied:
ok

roshan8 commented on cmd/provider_cmd_squadcast.go:
```diff
@@ -0,0 +1,45 @@
+		},
+	}
+	cmd.AddCommand(listCmd(newSquadcastProvider()))
+	baseProviderFlags(cmd.PersistentFlags(), &options, "user", "")
```
why do we need this?

mohammadmuazam replied:
It adds sample value of `--resources` flag (``terraformer import squadcast --help``)
![image](https://user-images.githubusercontent.com/110392771/187366313-6375968e-df0c-4deb-a620-1a4dbf44f11f.png)

roshan8 replied:
sure

roshan8 commented on docs/squadcast.md:
```diff
@@ -0,0 +1,68 @@
+
+Syntax:
+
+`export SQUADCAST_REFRESH_TOKEN=<YOUR_SQUADCAST_REFRESH_TOKEN>`
```
have a line stating refresh token can be passed as a flag or else as one can set the env

mohammadmuazam replied:
We have added this

roshan8 commented on docs/squadcast.md:
```diff
@@ -0,0 +1,68 @@
+    - slo
+    - runbook
+
+- `--region`
+  - Supported Values:
+    - `us`
+    - `eu`
```
fix the aligment

mohammadmuazam replied:
It looks something like this. 
![image](https://user-images.githubusercontent.com/110392771/187366627-403764ac-b37f-4715-88b8-6b1916bbd67e.png)

roshan8 commented on providers/squadcast/escalation_policy.go:
```diff
@@ -0,0 +1,80 @@
+)
+
+type EscalationPolicyGenerator struct {
+	SquadcastService
+	teamID string
```
Only one field is exported. let's either export both or none at all.

Lucifer8729 replied:
fixed after making use of `Request` function in `squadcast_service.go` file

roshan8 commented on providers/squadcast/runbook.go:
```diff
@@ -0,0 +1,79 @@
+
+type RunbookGenerator struct {
+	SquadcastService
+	teamID string
```
same comment applies everywhere

Lucifer8729 replied:
fixed after making use of `Request` function in `squadcast_service.go` file

roshan8 commented on providers/squadcast/team_roles.go:
```diff
@@ -0,0 +1,81 @@
+	escapedTeamName := url.QueryEscape(teamName)
+	getTeamURL := fmt.Sprintf("/v3/teams/by-name?name=%s",escapedTeamName)
+
+	team, err := g.generateRequest(getTeamURL)
```
we are fetching team name in all the resources. 
can we fetch them just once? may be in provider_cmd_squadcast.go

mohammadmuazam replied:
We have implemented this logic for both teamID and serviceID

roshan8 commented on providers/squadcast/squadcast_provider.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+const (
+	UserAgent = "terraform-provider-squadcast"
```
let's change this to `terraformer-squadcast`

mohammadmuazam replied:
Updated

roshan8 commented on providers/squadcast/squadcast_provider.go:
```diff
@@ -0,0 +1,179 @@
+	p.Service.SetVerbose(verbose)
+	p.Service.SetProviderName(p.GetName())
+	p.Service.SetArgs(map[string]interface{}{
+		"access_token":  p.accesstoken,
+		"refresh_token": p.refreshtoken,
```
only 1 needs to be set to be used in imported resources. please remove refresh token

mohammadmuazam replied:
This has been resolved

roshan8 commented on providers/squadcast/squadcast_provider.go:
```diff
@@ -0,0 +1,179 @@
+	return nil
+}
+
+func (p *SquadcastProvider) GetConfig() cty.Value {
```
please add one comment for this function and onemore while setting args.
`SetArgs` are used for fetching details within other files in the terraformer code.
`GetConfig` is used to send details to provider block of `terraform-provider-squadcast`

mohammadmuazam replied:
Added comments

roshan8 commented on providers/squadcast/squadcast_provider.go:
```diff
@@ -0,0 +1,179 @@
+	return map[string]interface{}{
+		"provider": map[string]interface{}{
+			"squadcast": map[string]interface{}{
+				"region": p.region,
```
just wondering, can't we set `source` block here itself? since it's just map[string]

Lucifer8729 replied:
we can't do that actually, it'll throw an error
![image](https://user-images.githubusercontent.com/72292326/187365811-5298223b-42ca-48b6-b5d8-4e4abbe5a34e.png)

Also https://www.terraform.io/language/providers/configuration#version-deprecated

roshan8 replied:
sure

Commit:
fix: fixed all suggested changes

Commit:
fix: merged request functions

Lucifer8729 committed:
fix: updated dependency from genereateRequest to Request function for all resources

Lucifer8729 committed:
fix: refactored code

replaced "sqaudcast" with GetProviderName function to follow the convention, used Sprintf, also removed not required attributes from different structs.
